From ad3e3da7ab6ea33b329a9f325286bad9fc7ea56a Mon Sep 17 00:00:00 2001
From: avatarAhmed Amine HAOUES <ahmed-amine.haoues_ext@softathome.com>
Date: Sun, 26 Oct 2025 19:18:35 +0530
Subject: [PATCH 1/1] pwhm: Fix hostapd crash due to invalid MLD context

PROVISORY: fix hostapd crash due to invalid MLD context


Problem: When updating hostapd parameters via wpactrl SET (e.g., setting
`mld_ap=1` and toggling `disable_11be`), there's a timing window before
the full configuration is reloaded where the internal state becomes
inconsistent.

Cause: Setting `mld_ap=1` causes hostapd to assume the BSS is part of
an MLD and that its MLD context is initialized. However, until the full
configuration is reloaded, this context is not yet valid. During this
window, hostapd may attempt to operate on an interface with `mld_ap=1`
but without a valid MLD context, leading to crashes.

Fix: To prevent this, we need to avoid updating hostapd memory if a
reload action is already scheduled in pwhm. This ensures consistency
between configuration state and runtime behavior.
In additon, make `mld_ap` and `disable_11be` set together.

GitOrigin-RevId: 33e1c216454f635d189d9bbf0cd9db8cffdf1c2e

ExternalID: SSW-11554

Fixes: PPW-1059

References: PPW-308, PCF-2145

Signed-off-by: default avatarAhmed Amine HAOUES <ahmed-amine.haoues_ext@softathome.com>
Signed-off-by: default avatarHoussem Dafdouf <houssem.dafdouf_ext@softathome.com>
Signed-off-by: Abhi Teja Cholleti <achollet@qti.qualcomm.com>

---
 src/nl80211/wld_hostapd_ap_api.c  | 22 +++++++++++++++++++---
 src/nl80211/wld_hostapd_cfgFile.c |  3 ++-
 2 files changed, 21 insertions(+), 4 deletions(-)

diff --git a/src/nl80211/wld_hostapd_ap_api.c b/src/nl80211/wld_hostapd_ap_api.c
index 5cc37cb..d40953f 100644
--- a/src/nl80211/wld_hostapd_ap_api.c
+++ b/src/nl80211/wld_hostapd_ap_api.c
@@ -382,14 +382,30 @@ swl_rc_ne wld_ap_hostapd_setParamAction(const char* paramName, wld_secDmn_action
 
 static bool s_setParam(T_AccessPoint* pAP, const char* param, const char* value, wld_secDmn_action_rc_ne* pAction) {
     ASSERTS_NOT_NULL(param, false, ME, "NULL");
-    bool ret = wld_ap_hostapd_setParamValue(pAP, param, value, param);
     wld_secDmn_action_rc_ne* pMappedAction = (wld_secDmn_action_rc_ne*) swl_table_getMatchingValue(&sHapdCfgParamsActionMap, 1, 0, param);
+    bool ret = true;
+
+    /* AP disabled (set the runtime value anyway) */
     if((pAP->status == APSTI_DISABLED) && !pAP->enable) {
         W_SWL_SETPTR(pAction, SECDMN_ACTION_OK_DONE);
-    } else if(pMappedAction != NULL) {
+	return wld_ap_hostapd_setParamValue(pAP, param, value, param);
+    }
+
+    /* set param with action (if needed SET runtime value) */
+    if(pMappedAction != NULL) {
+
         //keep most critical action
         W_SWL_SETPTR(pAction, SWL_MAX(*pAction, *pMappedAction));
-    } else if(ret) {
+	/* if within the action there is a conf file relaod, no need to set hostapd memory
+         * (reduce invalid conf window) */
+        if(*pAction < SECDMN_ACTION_OK_NEED_RESTART) {
+            ret = wld_ap_hostapd_setParamValue(pAP, param, value, param);
+        }
+        return ret;
+    }
+
+    ret = wld_ap_hostapd_setParamValue(pAP, param, value, param);
+    if(ret) {
         /*
          * If param is successfully set but has no specific applying action,
          * then reload whole save hostapd conf
diff --git a/src/nl80211/wld_hostapd_cfgFile.c b/src/nl80211/wld_hostapd_cfgFile.c
index a7cf883..7a5a8a8 100644
--- a/src/nl80211/wld_hostapd_cfgFile.c
+++ b/src/nl80211/wld_hostapd_cfgFile.c
@@ -812,10 +812,11 @@ static bool s_setVapCommonConfig(T_AccessPoint* pAP, swl_mapChar_t* vapConfigMap
              * 0 = no (no MLO)
              * 1 = yes (MLO) */
             swl_mapCharFmt_addValInt32(vapConfigMap, "mld_ap", 1);
-            enableVap11be = true;
+	    swl_mapCharFmt_addValInt32(vapConfigMap, "disable_11be", 0);
         } else {
             /* if no MLO, then no 11BE */
             swl_mapCharFmt_addValInt32(vapConfigMap, "disable_11be", 1);
+	    swl_mapCharFmt_addValInt32(vapConfigMap, "mld_ap", 0);
         }
     }
 
-- 
2.34.1

